libs/core/tests/yaml/gtest/document_tests.cpp

```diff
@@ -0,0 +1,686 @@
+  EXPECT_EQ(obj1.size(), 2);
+}
+
+TEST(YamlTests, Example2_10Test)
+{  // Example 2.10.  Node for “Sammy Sosa” appears twice in this document
+  char const *text = R"(---
+hr:
+  - Mark McGwire
+  # Following node labeled SS
+  - &SS Sammy Sosa
+rbi:
+  - *SS # Subsequent occurrence
+  - Ken Griffey)";
+
+  YamlDocument doc;
+  ASSERT_NO_THROW(doc.Parse(text));
+
+  auto const &root = doc.root();
+  EXPECT_TRUE(root.IsObject());
+  EXPECT_EQ(root.size(), 2);
+  EXPECT_TRUE(root.Has("hr"));
+  EXPECT_TRUE(root.Has("rbi"));
+
+  auto const &obj0 = root["hr"];
+  EXPECT_TRUE(obj0.IsArray());
+  EXPECT_EQ(obj0.size(), 2);
+  auto const &name0 = obj0[1];
+  EXPECT_EQ(name0.As<string>(), "Sammy Sosa");
+
+  auto const &obj1 = root["rbi"];
+  EXPECT_TRUE(obj1.IsArray());
+  EXPECT_EQ(obj1.size(), 2);
+  auto const &name1 = obj1[0];
+  EXPECT_EQ(name1.As<string>(), "Sammy Sosa");
+}
+
+TEST(YamlTests, Example2_12Test)
+{  // Example 2.12 Compact Nested Mapping
+  char const *text = R"(# Products purchased
+- item    : Super Hoop
+  quantity: 1
+- item    : Basketball
+  quantity: 4
+- item    : Big Shoes
+  quantity: 1
+)";
+
+  YamlDocument doc;
+  ASSERT_NO_THROW(doc.Parse(text));
+
+  auto const &root = doc.root();
+  EXPECT_TRUE(root.IsArray());
+  EXPECT_EQ(root.size(), 3);
+
+  auto const &obj0 = root[0];
+  EXPECT_TRUE(obj0.IsObject());
+  EXPECT_EQ(obj0.size(), 2);
+  auto const &name0 = obj0["item"];
+  EXPECT_EQ(name0.As<string>(), "Super Hoop");
+  auto const &quantity0 = obj0["quantity"];
+  EXPECT_EQ(quantity0.As<int>(), 1);
+
+  auto const &obj1 = root[1];
+  EXPECT_TRUE(obj1.IsObject());
+  EXPECT_EQ(obj1.size(), 2);
+  auto const &name1 = obj1["item"];
+  EXPECT_EQ(name1.As<string>(), "Basketball");
+  auto const &quantity1 = obj1["quantity"];
+  EXPECT_EQ(quantity1.As<int>(), 4);
+
+  auto const &obj2 = root[2];
+  EXPECT_TRUE(obj2.IsObject());
+  EXPECT_EQ(obj2.size(), 2);
+  auto const &name2 = obj2["item"];
+  EXPECT_EQ(name2.As<string>(), "Big Shoes");
+  auto const &quantity2 = obj2["quantity"];
+  EXPECT_EQ(quantity2.As<int>(), 1);
+}
+
+TEST(YamlTests, Example2_16Test)
+{  // Example 2.16.  Indentation determines scope
+  char const *text = R"(name: Mark McGwire
+accomplishment: >
+  Mark set a major league
+  home run record in 1998.
+stats: |
+  65 Home Runs
+  0.278 Batting Average)";
+
+  YamlDocument doc;
+  ASSERT_NO_THROW(doc.Parse(text));
+
+  auto const &root = doc.root();
+  EXPECT_TRUE(root.IsObject());
+  EXPECT_EQ(root.size(), 3);
+  EXPECT_TRUE(root.Has("name"));
+  EXPECT_TRUE(root.Has("accomplishment"));
+  EXPECT_TRUE(root.Has("stats"));
+
+  auto const &name = root["name"];
+  EXPECT_EQ(name.As<string>(), "Mark McGwire");
+
+  auto const &acc = root["accomplishment"];
+  EXPECT_EQ(acc.As<string>(), "Mark set a major league home run record in 1998.");
+
+  auto const &stats = root["stats"];
+  EXPECT_EQ(stats.As<string>(), "65 Home Runs\n0.278 Batting Average");
+}
+
+TEST(YamlTests, Example2_17Test)
```

Comment:
This test seems more comprenensive than yaml_tests and it is testing the same thing?

Reply:
No, they are different. `yaml_tests` tests the equivalence between YAML defined objects and JSON defined objects. This test is testing on a more fundamental basis the individual components.